File: Server/BaseBookingServices/models/BookingModel.js
import mongoose from 'mongoose';

const Schema = mongoose.Schema;

const BookingSchema = new Schema(
    {
        userId: {
            type: Schema.Types.ObjectId,
            required: false,
            ref: 'User',
        },
        driverId: {
            type: Schema.Types.ObjectId,
            required: true,
            ref: 'User',
        },
        customerName: {
            type: String,
            default: null,
        },
        customerPhone: {
            type: String,
            default: null,
        },
        type: {
            type: String,
            required: true,
            enum : ['Motorbike','Car'],
        },
        pickupLocation: {
            address: {
                type: String,
                required: true,
            },
            coordinate: {
                lat: { // x
                    type: Number,
                    required: true,
                },
                lng: { // y
                    type: Number,
                    required: true,
                },
            },
        },
        destinationLocation: {
            address: {
                type: String,
                required: true,
            },
            coordinate: {
                lat: { // x
                    type: Number,
                    required: true,
                },
                lng: { // y
                    type: Number,
                    required: true,
                },
            },
        },
        distance: {
            type: Number,
            required: true,
        },
        preTotal: {
            type: Number,
            required: true,
        },
        total: {
            type: Number,
            required: true,
        },
        promotionId: {
            type: Schema.Types.ObjectId,
            required: false,
            ref: 'Promotion',
        },
        status: {
            type: String,
            default: 'Pending',
            enum : ['Pending','Confirm','In-progress','Completed','Cancel','Pre-book']
        },
        pickupTime: {
            type: Date,
            default: Date.now(),
        },
        dropOffTime: {
            type: Date,
            default: Date.now(),
        },
        paymentMethodId: {
            type: Schema.Types.ObjectId,
            required: true,
            ref: 'PaymentMethod',
        },
        refundId: {
            type: Schema.Types.ObjectId,
            ref: 'Refund',
        },
    },
    {
        timestamps: true,
    });

export const Booking = mongoose.model('Booking', BookingSchema, "bookings");

File: Client/App/Driver/lib/global/models/booking/booking_model.dart
class BookingModel {
  final String userId;
  final String driverId;
  final String customerName;
  final String customerPhone;
  final String type;
  final String pickupAddress;
  final String destinationAddress;
  final String pickupTime;
  final String dropOffTime;
  final String createdAt;
  final String paymentMethodId;
  final String promotionId;
  final String status;
  final int total;
  final int distance;

  BookingModel(
    this.userId,
    this.driverId,
    this.customerName,
    this.customerPhone,
    this.type,
    this.pickupAddress,
    this.destinationAddress,
    this.pickupTime,
    this.dropOffTime,
    this.createdAt,
    this.paymentMethodId,
    this.promotionId,
    this.status,
    this.total,
    this.distance,
  );

  BookingModel.fromJson(Map<String, dynamic> json)
      : userId = json['userId'],
        driverId = json['driverId'],
        customerName = json['customerName'],
        customerPhone = json['customerPhone'],
        type = json['type'],
        pickupAddress = json['pickupLocation']['address'],
        destinationAddress = json['destinationLocation']['address'],
        pickupTime = json['pickupTime'],
        dropOffTime = json['dropOffTime'],
        createdAt = json['createdAt'],
        paymentMethodId = json['paymentMethodId'],
        promotionId = json['promotionId'],
        status = json['status'],
        total = json['total'],
        distance = json['distance'];
}
